### DISCUSSION TITLE
Color Correction: How to use spectral data?

### DISCUSSION THREAD

**Discussion by xeggbee** [colour-science/colour#976] [Q&A] [answered]
Hello,
first of all, this is a quite amazing lib with awesome documentation and examples! thanks for all, one of my favorite libs.

I'm looking for a help, My goal is to use the spectral reflectance of colorchecker in place of reference chart under controlled light  for color corrections.

Collect under noncontroled light, the reflectance of colorchecker, to use as reference in color correction. however, it donot work as expected, perhaps, i'm missing something.

this image has a resume of the approach:

![image](https://user-images.githubusercontent.com/39626787/167267961-cac41a0b-1b54-4d36-b973-3c0c33b0dd90.png)

here is the code:
This is the reflectance data read with spectralradiometer: (400nm to 700nm - 10nm interval) [ the reflectance data is in the botton]

### Converted the DNG to tiif

`with rawpy.imread(path) as raw:
    rgb = raw.postprocess(gamma=(1,1), no_auto_bright=False, output_bps=16)
imageio.imsave('linearf.tiff', rgb)`

### **I tried also this approach, used the spectral sensitivity of iphone, and got worst results!

`dicionario = [c1,c2,c3,c4,c5,c6,c7,c8,c9,c10,c11,c12,c13,c14,c15,c16,c17,c18,c19,c20,c21,c22,c23,c24]

illumini = np.array([0.36888,0.37556])

for i in dicionario:
    w = colour.SpectralShape(400, 700, 10).range()
    sd = colour.SpectralDistribution(i,w, name="Red")
    #sd = colour.SpectralDistribution(sdred)
    cmfs = colour.MSDS_CMFS['CIE 1931 2 Degree Standard Observer']
    #cmfs = mds
    illuminant = colour.SDS_ILLUMINANTS['FL2']
    # Calculating the sample spectral distribution *CIE XYZ* tristimulus values.
    XYZ = colour.sd_to_XYZ(sd, cmfs, illuminant)
    print("XYZ",XYZ)`

### this is the ouput, so adopted as REFERENCE:

`REFERENCE_SWATCHES = np.array(
[ 12.25241455 , 11.12983374 ,  8.05757363],
[ 38.68625739 , 35.57698923 , 25.97032945],
[ 19.14303218 , 20.37076879 , 37.93197895],
[ 11.84551613 , 14.84249584 ,  8.41315577],
[ 27.95345225 , 26.25257202 , 49.75300681],
[ 33.91225024 , 46.86674924 , 51.08620874],
[ 41.88449903 , 34.09754899 ,  7.8488979 ],
[ 16.58898647 , 14.44715502 , 48.02072235],
[ 33.33991078 , 22.69811628 , 16.85787418],
[  9.57883105 ,  7.43445209 , 16.47675857],
[ 36.50069127 , 49.02387214 , 14.35590841],
[ 49.71155283 , 47.64141496 , 10.60001274],
[ 10.58709859 ,  8.14387674 , 37.31081674],
[ 18.30880062 , 30.026641   , 13.9412999 ],
[ 23.52902061 , 14.75572623 ,  7.05016962],
[ 63.28089678 , 69.3177278  , 11.21161123],
[ 32.07133971 , 21.19445189 , 36.57583384],
[ 15.27999957 , 20.42215685 , 43.77232285],
[ 77.18688418 , 82.19742309 , 88.51959525],
[ 51.78219196 , 54.46986767 , 65.21142823],
[ 32.77697271 , 34.57889788 , 41.77492876],
[ 17.13248163 , 18.00805449 , 21.91202826],
[  8.14527274 ,  8.61479399 , 10.73871238],
[ 3.12459782,  3.24533775 , 4.2029096 ]
)`

### After used the colour checker dectetion lib, identified the the swatches, and applied the rREFERENCE above to correct the image.

`RGB = colour.read_image('/home/user1/phd/novafotos/t1/linearf.tiff')

REFERENCE_SWATCHES

##reshape the SWATCHES
SWATCHES
arr = np.array(SWATCHES)
arr.shape
newarr = arr.reshape(24,3)
newarr.shape

##CCM method
colour.characterisation.matrix_colour_correction_Finlayson2015(newarr, REFERENCE_SWATCHES) 

## to new Image
newimg = colour.colour_correction(RGB, newarr, REFERENCE_SWATCHES);

print ("Cor corrigida")
colour.plotting.plot_image(newimg);

## plot real image.
print ("Cor Original")
colour.plotting.plot_image(RGB);`

### the color correct output was:

![image](https://user-images.githubusercontent.com/39626787/167268614-97aa2ab7-341a-42ca-bd3f-13a5e18a721a.png)

### DNG Image is available here:

https://drive.google.com/drive/folders/1ebZYz8FceQDrMhEWAs1eYhB26dFpxUvk?usp=sharing

### reflectance data from colorcheck is available here:

`c1=  [0.0809165676664566,0.0806508711347853,0.0799413940390052,0.0710446859636388,0.0679562611110996,0.0740496564429403,0.0723114726794906,0.0746940377078699,0.0795947875197597,0.0761993044167605,0.0806278338083472,0.0801731955758011,0.0836462586924701,0.0845687269685092,0.0836891199598271,0.088104358919556,0.10423558212192,0.120814247660494,0.134033365633401,0.150046566716123,0.158236302605175,0.158079645595616,0.164159718705668,0.16799963758878,0.176384724099562,0.17897967493033,0.189306439846835,0.18575871491933,0.180204753253516,0.178735111068759,0.180513192367443,]
c2=  [0.190329341386665,0.18904986819075,0.188775393708127,0.198251728152685,0.206624760616321,0.216770115489317,0.231814304745027,0.266069066921539,0.295651985316632,0.332209740567,0.355592831571654,0.35779423844788,0.322752546499904,0.281027069914391,0.262014398282119,0.274122884024701,0.277684599112966,0.296829443329005,0.350308600649323,0.439076745091536,0.513744502459802,0.550067506935536,0.576990307838758,0.593526600444561,0.61502662047549,0.629875411419358,0.645313655685536,0.667696343049521,0.685106090350188,0.699773889324652,0.724704427194384,]
c3=  [0.269805033556456,0.317110758272039,0.312899352510869,0.343816846565163,0.353398731681145,0.370437336171116,0.366684423926127,0.358600983602075,0.346290298336726,0.313994438226565,0.297456353586563,0.270819206135311,0.255309488995945,0.221504291004738,0.201695957165123,0.196650366580751,0.186869448257657,0.172713222996244,0.163745864470564,0.161591005223112,0.160843605755638,0.14747088868927,0.143103578955389,0.142751434284373,0.155419091726337,0.162130168881458,0.171504542513289,0.179116250386558,0.180201749226625,0.171777799875648,0.16307381381121,]
c4=  [0.0732681358547101,0.0607446077139217,0.072627545022256,0.0658114997548549,0.0631869370041522,0.0687500273033089,0.0724362726807052,0.0832008745395024,0.0988839236725894,0.100093134618112,0.108419017269787,0.118903897849935,0.15637255205383,0.196571421052918,0.185107598053124,0.178053828139294,0.170186657787386,0.153848044151157,0.140193835474962,0.131871017363931,0.122066211449881,0.110239842613681,0.115119994428009,0.11607261708323,0.124188251968045,0.125445263117755,0.125237190684298,0.129419728274139,0.130897723758633,0.127098250788147,0.129270377493148,]
c5=  [0.398733838874891,0.471592795971482,0.468604921810866,0.473777585115652,0.473103565122924,0.4811031025931,0.468275080686624,0.456147090385259,0.427610977484268,0.395272531224525,0.390126138356314,0.347679335334133,0.302359005262652,0.280855594572316,0.254668651875885,0.234149989543298,0.204583190848587,0.173241700857504,0.160417572455386,0.177088092127742,0.215121009875599,0.266738462701649,0.31412021548474,0.34422121818684,0.409389619399585,0.456551138929139,0.503016449812413,0.503407610312948,0.485380939187423,0.476583285965494,0.485614557288132,]
c6=  [0.32283000734067,0.364989733466725,0.384466927197467,0.394740219942819,0.404181276988327,0.430658494840201,0.46853415701326,0.52770473206623,0.582101758574447,0.604163240006125,0.647303778756578,0.666807119654712,0.646058784106072,0.619508512382871,0.540828068140879,0.515193069807154,0.494755664146752,0.436075003111624,0.376719457283843,0.334516545966307,0.288945527024738,0.228130343649467,0.218315839275051,0.201886968537912,0.20594853763241,0.193386169295268,0.195059644642484,0.191614357028852,0.196842391494765,0.204281414732883,0.220401594674209,]
c7=  [0.0655098084472336,0.0492051451401083,0.0610595757388661,0.0648804225415474,0.0577339432693359,0.0635664740449796,0.0677469865682014,0.0781685615430193,0.0908563040456496,0.0924238602061924,0.103035722685506,0.11592767548847,0.151119531944832,0.184836314300633,0.181617871792142,0.209891897688008,0.30808329399749,0.429861695212618,0.520147811239192,0.596633182065675,0.64596364059829,0.614247275474633,0.65560225341857,0.648336115768061,0.686815965390762,0.672871527974438,0.687500875105106,0.693042213282444,0.685478048240984,0.690515421485792,0.68635607869496,]
c8=	 [0.339820583496756,0.409120344951529,0.442188797485009,0.459572103973604,0.464231198722582,0.495611624495967,0.479419772914532,0.451247857682308,0.389130079304461,0.30939271466162,0.260180528734269,0.214195590796208,0.18012297095046,0.15064199025092,0.115544381225992,0.109991355165261,0.113975940504094,0.111123757400069,0.0982214661676833,0.0986603717660494,0.099817344861785,0.0916320132844976,0.0994414797664642,0.0980098870752908,0.115418178403751,0.129143794595031,0.148599009920606,0.174036876056866,0.208443770550005,0.254737518775746,0.317335303173492,]
c9=  [0.196543206763699,0.179588945092157,0.186525140202341,0.160698238923036,0.153858689209312,0.160023263604523,0.153526826324801,0.149029758142114,0.145827670290368,0.135713434249522,0.135712934303049,0.137380328215663,0.12869751924122,0.116933924606016,0.102924987417334,0.103653601355773,0.118549777381867,0.13812042013082,0.184866524322182,0.323119944575716,0.496601480733114,0.562095192764694,0.663547811105562,0.678191108051094,0.743839611221814,0.724522043318302,0.743244443050635,0.743961480963106,0.744885344361183,0.735052707577471,0.741864141321583,]
c10= [0.210442255413856,0.202365404128712,0.210281516212622,0.180525564265681,0.168294042789932,0.163449213039288,0.147379632924175,0.132643338701665,0.113325125017658,0.0976594401470317,0.0942297906753103,0.0841916134623119,0.0721605810907962,0.063862688223389,0.0572893947812231,0.0542626799525423,0.0544365573747611,0.0566650763920967,0.0528372376628998,0.0559154604368737,0.0675048871459579,0.0974315086354773,0.123655703313483,0.133371087548342,0.152694422430528,0.18330642348231,0.2137361134525,0.234984118513444,0.250294917455561,0.255048493458923,0.250233078303442,]
c11= [0.0898178710172041,0.0943439392409357,0.0987160133886949,0.0897441526438344,0.0921473797769371,0.100161662953095,0.111210805854335,0.125551583292987,0.1694682063199,0.216027965689057,0.326337053717786,0.464866974419545,0.575564583250627,0.629835514609691,0.584258988873277,0.579125343078094,0.59855420184191,0.553276126484279,0.498152369943084,0.451568283851573,0.404509618906246,0.347805449360627,0.348903791281311,0.338215961878453,0.346816275887952,0.335834215097049,0.345740223809102,0.360442347865578,0.378693362061194,0.391179483143634,0.403385687522362,]
c12= [0.101769421076977,0.101485227250533,0.0909768687718493,0.0855626479129179,0.0799981890037773,0.0829608981486767,0.087522675293731,0.0936609622205263,0.11001535984835,0.116906635468068,0.158143486875959,0.212894327768237,0.289707291489185,0.395301628130759,0.405437284001913,0.439834177130859,0.538798560184266,0.599094619493414,0.621543077717915,0.658759818161822,0.688540594469187,0.650004788380835,0.692788808448299,0.686670921081225,0.730325474874111,0.718988306309388,0.734567774413891,0.733256657252528,0.731175844252551,0.732727336002906,0.735833467914285,]
c13= [0.261497974632618,0.271936976751908,0.314457472728486,0.332391874831977,0.350790505703692,0.388881123182652,0.394394099426998,0.377235102531002,0.31906526623327,0.227073287891533,0.169635767407625,0.122695611796528,0.0919680234301705,0.0744714044699647,0.0546848334846939,0.0545496443426618,0.0591725633536876,0.0560166338492272,0.0534055730712742,0.049481071410316,0.0492252076657012,0.0440533584281579,0.0515067383166855,0.0508435245070514,0.0579355121770477,0.0620442347902271,0.0645184072806307,0.0682065714372249,0.0767610531016927,0.0851291691213357,0.101834496410224,]
c14= [0.0955393898130529,0.0894419266841759,0.0885226074660729,0.0894717433574231,0.0852209076231497,0.0951681786115778,0.111020731937353,0.134649684535393,0.176024542349483,0.218154552591454,0.319789676559454,0.42407778039297,0.469369310159964,0.462497231221654,0.364932575141951,0.347776845019802,0.356386244369832,0.310498116153074,0.243667387746917,0.193021841930964,0.160338488402015,0.115101065247139,0.121318975161964,0.11231319678338,0.127389178230738,0.113926785409379,0.121746900847101,0.12524359860488,0.137150329017568,0.142662199223682,0.155008402090936,]
c15= [0.0671644190518135,0.0709552699997325,0.0681960473173244,0.0577064765129307,0.0550597230604484,0.0628263479032059,0.0645363918303289,0.0705433922154122,0.0753205164660728,0.0685980854068525,0.0757424184679,0.0777999207022899,0.0716645565486953,0.0724628889418917,0.0577216779731495,0.0573343384069645,0.0726706778876113,0.0792867983899378,0.0930395695186142,0.143635497018084,0.260843952876956,0.380842277449101,0.554137422688026,0.62929441302786,0.731725380134868,0.729032330800507,0.754326389050919,0.762723952606923,0.765250311964308,0.759121298943218,0.761120676698035,]
c16= [0.0989470361317674,0.0789413014178204,0.0798468647045374,0.0764365096614674,0.0700135025381538,0.0799281626617584,0.0809596610730403,0.0950951244149038,0.112078300232436,0.134819887522751,0.240514992521544,0.422470945416702,0.596445731931632,0.718501940512704,0.69880699065533,0.734722894889699,0.832055002963157,0.84879440956655,0.819648978445186,0.822432738102604,0.828107277246127,0.76364159671261,0.802782569220548,0.789473040357827,0.83463610852822,0.81659375363963,0.832062167432181,0.832330073426954,0.837581352048969,0.838024028808176,0.836428663080232,]
c17= [0.383589574346733,0.431818845252891,0.444151712526414,0.416321174634962,0.38382538962536,0.357726501187422,0.327803713693065,0.29454894671327,0.254821203640706,0.21540663627425,0.195033248991343,0.175937993199296,0.147965230878031,0.115443959660776,0.100406648507782,0.104090743296691,0.114444897226124,0.123323496445216,0.158609862486889,0.243137446837478,0.360360234059175,0.47163579361669,0.538043919131857,0.576176750621855,0.626989229799632,0.695723830534721,0.745022692880205,0.779362422214134,0.801368984410729,0.801013905443247,0.802994379972699,]
c18= [0.244788886054003,0.284260962032819,0.318184913530548,0.33422845059833,0.353478306406149,0.396551691723851,0.426217780062279,0.46822580263077,0.491690533622688,0.476991885316886,0.453288026657009,0.40663831626565,0.347851729926283,0.288683726354602,0.227708385263668,0.187569918372718,0.145476632699573,0.112864123054474,0.0978703845677431,0.0883645819097448,0.0791219146773573,0.0773982244611454,0.0729304516637471,0.0716547315914966,0.0688643882678333,0.0701544298089268,0.0720171819465011,0.0763688181346178,0.0691923575529477,0.06528461459588,0.0661981937382557,]
c19= [0.488558258772534,0.679739386787103,0.797599678623456,0.808469107885554,0.802674463194837,0.806542133350057,0.81806265363265,0.830967622352032,0.839249487191543,0.853009405725514,0.83717111647963,0.8260824892963,0.818056346090331,0.809648975622822,0.879153961411898,0.856983453453604,0.787716767796018,0.775951500780546,0.818038066880226,0.819359361008165,0.801649252061416,0.888159276686427,0.812573409679312,0.820106148125545,0.73197759203891,0.76119158338819,0.729744799379206,0.711029324528706,0.709746230545148,0.700091334396959,0.67603785182241,]
c20= [0.477695466513189,0.586601009843247,0.619525932039492,0.620037328917475,0.610531072802745,0.607898597700771,0.598741590930076,0.584455524214219,0.586246595131041,0.583772768954965,0.567062819859503,0.547281831777896,0.548520967275968,0.53529858702676,0.586145544711241,0.570988550872962,0.520676263106561,0.509846856164396,0.536398029810859,0.535056156732874,0.520866308913142,0.57810218849439,0.525997795862763,0.53099245617144,0.46923876807505,0.491298088057099,0.469371233600789,0.459076305712322,0.45577985754184,0.443098105110805,0.427985927906828,]
c21= [0.354956421720784,0.387773820037598,0.40268027464108,0.396464506110916,0.390078554083146,0.389674023421161,0.379476854022559,0.380868796152713,0.37104708689337,0.366720004456255,0.359890913255566,0.357037062466932,0.3466756986828,0.34404204713916,0.368620423230612,0.359579961330504,0.333577257110007,0.329699966330026,0.341247219891019,0.337368421258752,0.330064822366181,0.357391999266046,0.330663583538059,0.328657326191994,0.29396484470199,0.304778448541063,0.294068317550931,0.282935986603976,0.283301820308277,0.277128483990304,0.273514600352077,]
c22= [0.1935158851253,0.198272247438733,0.208419228383539,0.206049298698106,0.206012974433646,0.208694774282582,0.19989528068629,0.196609153364737,0.194233424402793,0.190713944386992,0.182881280204934,0.183596929903362,0.182312657640749,0.176681401354308,0.190018438289436,0.184426844823453,0.178115009633041,0.174415973358516,0.178673906199316,0.177722908043468,0.172260725840988,0.185874850947541,0.171638392279354,0.170804697494451,0.151749298477654,0.157618385782687,0.149676703278448,0.147584050574133,0.143281340996075,0.141304521044664,0.140231395986744,]
c23= [0.121029439390052,0.109773888035438,0.100667814334709,0.100388880590726,0.102216206233633,0.100314474160502,0.0972845788452191,0.0956349895782304,0.0985385811680834,0.0900413254570052,0.0909762953889136,0.0884382628297641,0.0926635017252373,0.0873946408891542,0.091144422380078,0.0877679599758725,0.0840049249914235,0.0803284846728752,0.0834123012319875,0.0834224127668517,0.0811703564204933,0.0866246885431312,0.0810484170948454,0.080859645605934,0.0713809812023999,0.0755128847709437,0.0734384723165986,0.0743153174345048,0.0675258209822046,0.0700620956868183,0.0659941985588062,]
c24= [0.0561036680174321,0.051537874435454,0.0458080637167191,0.0423227202907465,0.0409785901165612,0.0381342451151333,0.0362686678938866,0.0355105039176247,0.0354670632982508,0.0360875599090504,0.0354198728329977,0.0364499059881475,0.0295610883324002,0.0334870344860978,0.0329739834488695,0.0319293043018508,0.032617185707168,0.0318952614170191,0.0322102777986658,0.0326141669215172,0.0288281791203202,0.0321753954622247,0.0310495936693493,0.0290831986543141,0.031123601930286,0.0296012440877328,0.0295642313430246,0.0328110932013253,0.0263479258653895,0.0268351268208579,0.0322175799435598,]
`

**Comment by thomasmansencal** [comment, marked as answer]
Hi @xeggbee,

When you generate the reference values and perform the Spectral integration with the `colour.sd_to_XYZ` definition, you should try to divide by 100 because the values returned by default are expressed between 0 and 100 as percentage in the reflectance case.

Cheers,

Thomas

**Reply by xeggbee** [reply]
@KelSolaar 
### thanks! thanks thanks! it really changed!

![image](https://user-images.githubusercontent.com/39626787/167298620-2d14ecb1-f3a7-485f-aab3-dad85315ab84.png)

sounds like that the red is impacted.

![image](https://user-images.githubusercontent.com/39626787/167298646-920ee7f2-f6d2-49e4-a1af-ccf82fb05643.png)

i'll try to iterate over each illuminant and see if changes, i'm thinking like this because if I perform the color correction (ignoring the spectral data) using the standard D65, performs much better in the color correction.


![image](https://user-images.githubusercontent.com/39626787/167298692-c7d3f4f7-ff84-4cc6-9c73-78e0a1b6ca53.png)

![image](https://user-images.githubusercontent.com/39626787/167298740-73dc6cf9-39f6-4a2c-9b43-53721a61cb76.png)


### as next step, i'll try the steps below:
So, in poor understanding, the spectral correction might be impacted due the illuminant, due the DNG to tiff conversion (gamma and 8 or 16 bits output) or camera white balance (AWB, Tugsten, Fluorescent, Daylight and Cloudy). Please let know if have some suggestion over it. 

others concers:
### Where is the moment that should use the Spectral Sensitivity Function?
Another point is, for this flow, where is the moment that i can use the Spectral Sensitivy function (SSF) of camera? is during the sd_to_XYZ in place of CMFS?

`for i in dicionario:
    i = i/100
    w = colour.SpectralShape(400, 700, 10).range()
    sd = colour.SpectralDistribution(i,w, name="Red")
        #sd = colour.SpectralDistribution(sdred)
    cmfs = colour.MSDS_CMFS['CIE 1931 2 Degree Standard Observer'] ##replace by SSF?
    illuminant = colour.SDS_ILLUMINANTS['D60']
    # Calculating the sample spectral distribution *CIE XYZ* tristimulus values.
    XYZ = colour.sd_to_XYZ(sd, cmfs, illuminant)
    print("XYZ",XYZ)`


### Scene and CC reflectance.
Checked your comment in [#747747#discussioncomment-287904_](https://github.com/colour-science/colour/discussions/747#discussioncomment-287904) and you suggest to also capture the spectral irradiance of the scene. 
[naive question] How to capture it? should use a spectralradiometer(or whcih equipment?) and point it to the sky or the light source and export as irradiance? Does it go inplace of "SDS_ILLUMINANTS", and also have to generate the xy of CCS_Illuminans?

**for this data collection used the spectral radiometer Field Spec 4.

Also reported to use the spectro-photometer to capture spectral reflectance of the ColorChecker, my question here is, for case of non standard light source, does the spectro-photometer also works? (its much cheapper than spectralradiometer)
i'm asking because spectral-photometer needs generally to specify the light source, which cover just the standard CIE.


 spectral irradiance of the scene and the spectral reflectance of the ColorChecker Classic (with a spectro-photometer)

**Reply by thomasmansencal** [reply]
> Where is the moment that should use the Spectral Sensitivity Function?

If you are using a chart based workflow, you don't really need to use camera sensitivities, those do not require a chart because you can generate reference values at will and at any desired white-balance. We have a lot of relevant code for that here: https://github.com/colour-science/colour/blob/develop/colour/characterisation/aces_it.py and it is being used there for example: https://github.com/ampas/idt-calculator/

> How to capture it?

It depends on your device but you would probably either want a cosine corrector to sample the hemisphere above the chart or alternatively, measure a reflectance standard such as Spectralon where your chart is located. You could then use it to generate references values to match for your chart.

**Reply by xeggbee** [reply]
thanks again!

### Tried the Aces, but not sure where to go now.

`sensitivities = sds_and_msds_to_msds(read_sds_from_csv_file(path).values())
illuminant = SDS_ILLUMINANTS['D55']
B, b = matrix_idt(sensitivities, illuminant)
RGB = colour.read_image('/home/user1/phd/novafotos/t1/linear5.tiff')
camera_RGB_to_ACES2065_1(RGB, B, b)
`
[[ 0.70516773  0.18084909  0.07795802]
 [ 0.00535485  1.09510291 -0.11779157]
 [ 0.03451185 -0.35059552  1.28606089]]

[ 2.15243805  1.          1.81062535]

#plot got ping
colour.plotting.plot_image(camera_RGB_to_ACES2065_1(RGB, B, b));

![image](https://user-images.githubusercontent.com/39626787/169457391-963727fb-dc8a-4a2e-8b02-dc43a63fb560.png)

### how to move to other white-balance?

### how to export back to sRGB?

**Comment by xeggbee** [comment]
### Am I placing the Spectral Referece at the right place? [using the example from colour-checker-detection]
Delta E seems pretty good to be true!

## This is the SD to XYZ

`XYZ1= np.array([
 [ 0.11549559  ,0.10200941  ,0.06013436],
 [ 0.42861565  ,0.36950352  ,0.21554453],
 [ 0.17278399  ,0.18569993  ,0.29463422],
 [ 0.11139152  ,0.13672005  ,0.06425129],
 [ 0.24767728  ,0.23089142  ,0.37233453],
 [ 0.31010971  ,0.42671906  ,0.40115312],
 [ 0.40194304  ,0.31759328  ,0.05984117],
 [ 0.14358259  ,0.12888192  ,0.36174337],
 [ 0.31502616  ,0.20954881  ,0.12391923],
 [ 0.09363123  ,0.07044418  ,0.13090358],
 [ 0.37227452  ,0.48193868  ,0.11319805],
 [ 0.50611563  ,0.46548459  ,0.07672176],
 [ 0.08517642  ,0.06801938  ,0.2707255 ],
 [ 0.15638051  ,0.24885341  ,0.09747058],
 [ 0.21272349  ,0.12891096  ,0.04849688],
 [ 0.62043362  ,0.64889978  ,0.08697529],
 [ 0.32490704  ,0.21228402  ,0.27924422],
 [ 0.14687085  ,0.20296159  ,0.3653085 ],
 [ 0.89405879  ,0.94030569  ,0.83409987],
 [ 0.59408958  ,0.62559952  ,0.5815818 ],
 [ 0.36606854  ,0.38680561  ,0.36486591],
 [ 0.1901647   ,0.20015258  ,0.18977891],
 [ 0.08915284  ,0.0942116   ,0.09176753],
 [ 0.03240577  ,0.03395494  ,0.03266119]
])`

### In place of REFERENCE_SWATCHES, used the XYZ1 (the SD to XYZ)

`for i, swatches in enumerate(SWATCHES):
    swatches_xyY = colour.XYZ_to_xyY(colour.RGB_to_XYZ(
        swatches, D65, D65, colour.RGB_COLOURSPACES['sRGB'].matrix_RGB_to_XYZ))

    colour_checker = colour.characterisation.ColourChecker(
        os.path.basename(COLOUR_CHECKER_IMAGE_PATHS[i]),
        dict(zip(REFERENCE_COLOUR_CHECKER.data.keys(), swatches_xyY)),
        D65)
    
    colour.plotting.plot_multi_colour_checkers(
        [REFERENCE_COLOUR_CHECKER, colour_checker])
    xs = swatches **<--------** 
    swatches_f = colour.colour_correction(swatches, swatches, XYZ1) **<--------** 
    swatches_f_xyY = colour.XYZ_to_xyY(colour.RGB_to_XYZ(
        swatches_f, D65, D65, colour.RGB_COLOURSPACES['sRGB'].matrix_RGB_to_XYZ))
    colour_checker = colour.characterisation.ColourChecker(
        '{0} - CC'.format(os.path.basename(COLOUR_CHECKER_IMAGE_PATHS[i])),
        dict(zip(REFERENCE_COLOUR_CHECKER.data.keys(), swatches_f_xyY)),
        D65)
    
    colour.plotting.plot_multi_colour_checkers(
        [REFERENCE_COLOUR_CHECKER, colour_checker])

    colour.plotting.plot_image(colour.cctf_encoding(
        colour.colour_correction(
            COLOUR_CHECKER_IMAGES[i], swatches, XYZ1))); **<--------** 
`

`###Before Correction
print ("Before CCM - delta E")
real = XYZ_to_Lab(xs) 
cc = XYZ_to_Lab(XYZ1)  **<--------** 
print (delta_E(real , cc, method='CIE 2000')  )
print (np.sum(delta_E(real , cc, method='CIE 2000')))

##dAfter COrrection
print ("After CCM - delta E")
corr = XYZ_to_Lab(swatches_f) 
print (delta_E(corr, cc, method='CIE 2000'))  
print (np.sum(delta_E(corr, cc, method='CIE 2000')))`



![image](https://user-images.githubusercontent.com/39626787/169454976-00db6e14-70ac-4fed-b8d2-1b6f9dd38419.png)

![image](https://user-images.githubusercontent.com/39626787/169454998-14e17919-b563-4605-aacd-9ed329dd8c7f.png)


Before CCM - delta E
[ 24.88903759  30.48865173  14.31711188  11.81056074  16.56912886
  18.54136263  39.5667104   12.14440652  26.96516623  16.40594934
  10.65338232  38.69993748  18.3030553   16.12876003  29.26348574
  22.82370674  22.6653305   28.03323937  10.39938     24.24462958
  22.82424673  20.84249434  17.16353863  14.57846023]
**toltal sum: 508.321732928**

After CCM - delta E
[ 2.90629512  2.59789819  3.24360788  4.5494967   3.1298541   4.08681839
  0.83428088  2.26238368  1.17373848  3.57508397  3.85485893  1.00052719
  3.42749492  5.25820761  2.81237257  3.25988869  1.31852124  4.92900922
  7.60066046  2.8074503   2.73694856  3.04011806  3.28293047  4.49887239]
**toltal sum: 78.1873180067**


### Some suggestion to determine the SDS_ILLUMINANTS when capture was in non-controlled light?
I'm trying two approaches:
1-best_illuminat
--seek for the best_illumint wb that are the closest to the whitebalance from DNG, then use the suggested illumiant

2-delta E
--iterate over illuminants, and after the color correction check for the smalled illuminant.
